examples/app/headless_renderer.rs

```diff
@@ -0,0 +1,444 @@
+//! This example illustrates how to make headless renderer
+
+mod frame_capture {
+    pub mod image_copy {
+        use bevy::prelude::*;
+        use bevy::render::{
+            render_asset::RenderAssets,
+            render_graph::{self, NodeRunError, RenderGraph, RenderGraphContext},
+            render_resource::{
+                Buffer, BufferDescriptor, BufferUsages, CommandEncoderDescriptor, Extent3d,
+                ImageCopyBuffer, ImageDataLayout, Maintain, MapMode,
+            },
+            renderer::{RenderContext, RenderDevice, RenderQueue},
+            Extract, RenderApp,
+        };
+        use std::sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc,
+        };
+
+        pub fn receive_images(
+            image_copiers: Query<&ImageCopier>,
+            mut images: ResMut<Assets<Image>>,
+            render_device: Res<RenderDevice>,
+        ) {
+            for image_copier in image_copiers.iter() {
+                if !image_copier.enabled() {
+                    continue;
+                }
+                // Derived from: https://sotrh.github.io/learn-wgpu/showcase/windowless/#a-triangle-without-a-window
+                // We need to scope the mapping variables so that we can
+                // unmap the buffer
+                futures_lite::future::block_on(async {
+                    let buffer_slice = image_copier.buffer.slice(..);
+
+                    // NOTE: We have to create the mapping THEN device.poll() before await
+                    // the future. Otherwise the application will freeze.
+                    let (tx, rx) = async_channel::bounded(1);
+                    buffer_slice.map_async(MapMode::Read, move |result| {
+                        tx.send_blocking(result).unwrap();
+                    });
+                    render_device.poll(Maintain::Wait);
+                    rx.recv().await.unwrap().unwrap();
+                    if let Some(image) = images.get_mut(&image_copier.dst_image) {
+                        image.data = buffer_slice.get_mapped_range().to_vec();
+                    }
+
+                    image_copier.buffer.unmap();
+                });
```

Comment:
You don't need a future that blocks here. You can copy what is done in the gpu_readback example by using crossbeam_channel using `rx.recv()` that blocks. It makes the code a bit simple in my opinion.

Reply:
link to example I mentioned https://github.com/bevyengine/bevy/blob/main/examples/shader/gpu_readback.rs